src/ring_theory/graded_algebra/homogeneous_ideal.lean

```diff
@@ -0,0 +1,360 @@
+/-
+Copyright (c) 2021 Jujian Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jujian Zhang, Eric Wieser
+-/
+
+import ring_theory.ideal.basic
+import ring_theory.ideal.operations
+import linear_algebra.finsupp
+import ring_theory.graded_algebra.basic
+
+/-!
+
+# Homogeneous ideal of a graded algebra
+
+This file defines homogeneous ideals of `graded_algebra A` where `A : ι → submodule R R`and
+operations on them:
+* `mul`, `inf`, `Inf` of homogeneous ideals are homogeneous;
+* `⊤`, `⊥`, i.e. the trivial ring and `R` are homogeneous;
+* `radical` of a homogeneous ideal is homogeneous.
+-/
+
+noncomputable theory
+
+
+variables (R : Type*) [comm_ring R] {ι : Type*} [deq : decidable_eq ι]
+variables (A : ι → submodule R R)
+
+open_locale direct_sum pointwise big_operators
+open set direct_sum set_like
+
+section defs
+
+/--A homogeneous ideal is an `I : ideal R` such that `I` is generated by
+homogeneous elements in `I`.-/
+def is_homogeneous_ideal' (I : ideal R) : Prop :=
+  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x}
+
+/-- Given any ideal `I : R`, there is a homogeneous ideal generated by
+the homogeneous elements of `I`-/
+def homogeneous_ideal_of_ideal (I : ideal R) :
+  ideal R := ideal.span (set_of (is_homogeneous A) ∩ I)
+
+lemma homogeneous_ideal_of_ideal_le_ideal
+  (I : ideal R) :
+  homogeneous_ideal_of_ideal R A I ≤ I :=
+begin
+  rw homogeneous_ideal_of_ideal,
+  conv_rhs { rw ←ideal.span_eq I },
+  apply ideal.span_mono, exact (set_of (is_homogeneous A)).inter_subset_right ↑I,
+end
+
+variables [add_comm_monoid ι] [ga : graded_algebra A]
+include ga
+/--A homogeneous ideal is an `I : ideal R` such that `I` is generated by a set of
+homogeneous elements.-/
+def is_homogeneous_ideal (I : ideal R) : Prop :=
+  ∃ S : set (homogeneous_submonoid A), I = ideal.span (coe '' S)
+
+/--Equivalently, an `I : ideal R` is homogeneous iff for every element `r : I` and `i : ι`, the
+degree `i` component of `r` is in `I`.-/
+def is_homogeneous_ideal'' (I : ideal R) : Prop :=
+  ∀ r : R, r ∈ I → ∀ i : ι, graded_algebra.proj A i r ∈ I
```

Comment:
we don't actually want the three different definitions here:
```lean
def is_homogeneous_ideal (I : ideal R) : Prop :=
  ∃ S : set (homogeneous_submonoid A), I = ideal.span (coe '' S)

def is_homogeneous_ideal' (I : ideal R) : Prop :=
  I = ideal.span {x | x ∈ I ∧ is_homogeneous A x}

def is_homogeneous_ideal'' (I : ideal R) : Prop :=
  ∀ r : R, r ∈ I → ∀ i : ι, graded_algebra.proj A i r ∈ I
```
We should pick just one of these, and then define lemmas like:
```lean 
lemma is_homogeneous_ideal_iff_exists :
  is_homogeneous_ideal R A I ↔ ∃ S : set (homogeneous_submonoid A), I = ideal.span (coe '' S)
lemma is_homogeneous_ideal_iff_eq :
  is_homogeneous_ideal R A I ↔ I = ideal.span {x | x ∈ I ∧ is_homogeneous A x}
lemma is_homogeneous_ideal_iff_forall :
  is_homogeneous_ideal R A I ↔ ∀ r : R, r ∈ I → ∀ i : ι, graded_algebra.proj A i r ∈ I
```
My hunch is that `∀ r : R, r ∈ I → ∀ i : ι, graded_algebra.proj A i r ∈ I` is probably the best to use as the deifnition.

Reply:
In fact, I think this is probably the best definition to use:
```
def is_homogeneous_ideal (I : ideal R) : Prop :=
  ∀ (i : ι) {r : R}, r ∈ I → graded_algebra.proj A i r ∈ I

lemma is_homogeneous_ideal_iff_subset (I : ideal R) :
  is_homogeneous_ideal R A I ↔ ∀ i, (I : set R) ⊆ graded_algebra.proj A i ⁻¹' I :=
iff.rfl
```

Reply:
I am not sure, because `decompose A r i` and `I` doesn't `has_mem`, so one needs to cast.
```lean
def is_homogeneous_ideal (I : ideal R) : Prop :=
∀ r : R, r ∈ I → ∀ i : ι, (graded_algebra.decompose A r i : R) ∈ I
```

Reply:
Right, that cast is there above - see my edited post.

Reply:
Sorry for the late reply. I have applied your suggestion. Thank you.